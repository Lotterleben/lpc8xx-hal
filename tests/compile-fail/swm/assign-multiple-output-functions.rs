use lpc8xx_hal::{
    Peripherals,
    pins::{
        self,
        Pin,
    },
    swm,
};


fn main() {
    let mut p = Peripherals::take().unwrap();

    let     swm    = p.SWM.split();
    let mut syscon = p.SYSCON.split();

    #[cfg(feature = "82x")]
    let mut swm_handle = swm.handle;
    #[cfg(feature = "845")]
    let mut swm_handle = swm.handle.enable(&mut syscon.handle);

    let pio0_0: Pin<_, pins::state::Unused> = p.pins.pio0_0;

    let u0_rxd: swm::Function<_, swm::state::Unassigned> =
        swm.movable_functions.u0_rxd;
    let u0_txd: swm::Function<_, swm::state::Unassigned> =
        swm.movable_functions.u0_txd;
    let u1_rxd: swm::Function<_, swm::state::Unassigned> =
        swm.movable_functions.u1_rxd;
    let u1_txd: swm::Function<_, swm::state::Unassigned> =
        swm.movable_functions.u1_txd;

    let (u0_rxd, pio0_0) =
        u0_rxd.assign(pio0_0.into_swm_pin(), &mut swm_handle);
    let (u1_rxd, pio0_0) =
        u1_rxd.assign(pio0_0, &mut swm_handle);
    let (u0_txd, pio0_0) =
        u0_txd.assign(pio0_0, &mut swm_handle);
    // Should fail: Another output function already assigned to pin.
    let (u1_txd, pio0_0) =
        u1_txd.assign(pio0_0, &mut swm_handle);
}
