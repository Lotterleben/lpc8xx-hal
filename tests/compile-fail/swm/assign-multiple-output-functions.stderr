error[E0277]: the trait bound `lpc8xx_hal::pins::Pin<lpc8xx_hal::pins::PIO0_0, lpc8xx_hal::pins::state::Swm<((),), (((),),)>>: lpc8xx_hal::swm::AssignFunction<lpc8xx_hal::swm::U1_TXD, lpc8xx_hal::swm::Output>` is not satisfied
  --> $DIR/assign-multiple-output-functions.rs:41:23
   |
41 |         u1_txd.assign(pio0_0, &mut swm_handle);
   |                       ^^^^^^ the trait `lpc8xx_hal::swm::AssignFunction<lpc8xx_hal::swm::U1_TXD, lpc8xx_hal::swm::Output>` is not implemented for `lpc8xx_hal::pins::Pin<lpc8xx_hal::pins::PIO0_0, lpc8xx_hal::pins::state::Swm<((),), (((),),)>>`
   |
   = help: the following implementations were found:
             <lpc8xx_hal::pins::Pin<T, lpc8xx_hal::pins::state::Swm<(), ()>> as lpc8xx_hal::swm::AssignFunction<F, lpc8xx_hal::swm::Analog>>
             <lpc8xx_hal::pins::Pin<T, lpc8xx_hal::pins::state::Swm<(), Is>> as lpc8xx_hal::swm::AssignFunction<F, lpc8xx_hal::swm::Output>>
             <lpc8xx_hal::pins::Pin<T, lpc8xx_hal::pins::state::Swm<O, Is>> as lpc8xx_hal::swm::AssignFunction<F, lpc8xx_hal::swm::Input>>
